File: test_max_usage_fix.py
#!/usr/bin/env python3
"""
Test script to verify that the coupon max usage fix is working correctly.
This script tests both coupons that should be blocked and ones that should work.
"""

import asyncio
import motor.motor_asyncio
from datetime import datetime
import requests
import json
import logging

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

async def test_coupon_max_usage_fix():
    """Test the coupon max usage validation fix"""
    
    try:
        # Connect to MongoDB
        client = motor.motor_asyncio.AsyncIOMotorClient("mongodb://localhost:27017")
        
        # Access admin database where coupons are stored
        admin_db = client.get_database("admin")
        coupons_collection = admin_db.get_collection("coupons")
        
        # Access main database for orders
        main_db = client.get_database("MonkeyZ")
        orders_collection = main_db.get_collection("orders")
        
        print("🧪 TESTING: Coupon Max Usage Fix")
        print("=" * 50)
        
        # Test scenarios
        test_scenarios = [
            {
                'name': 'Test Coupon with Max Usage Exceeded',
                'coupon_code': 'TEST_MAX_EXCEEDED',
                'max_uses': 2,
                'should_be_blocked': True
            },
            {
                'name': 'Test Coupon with Max Usage Available', 
                'coupon_code': 'TEST_MAX_AVAILABLE',
                'max_uses': 10,
                'should_be_blocked': False
            },
            {
                'name': 'Test Coupon with Unlimited Usage',
                'coupon_code': 'TEST_UNLIMITED',
                'max_uses': None,
                'should_be_blocked': False
            }
        ]
        
        # First, let's check what real coupons exist and their status
        print("\n📋 EXISTING COUPONS STATUS:")
        print("-" * 30)
        
        active_coupons = await coupons_collection.find({'active': True}).to_list(None)
        for coupon in active_coupons:
            code = coupon.get('code')
            max_uses = coupon.get('maxUses')
            usage_count = coupon.get('usageCount', 0)
            
            # Count real orders
            real_count = await orders_collection.count_documents({
                '$or': [
                    {'couponCode': {'$regex': f'^{code}$', '$options': 'i'}},
                    {'coupon_code': {'$regex': f'^{code}$', '$options': 'i'}}
                ],
                'status': {'$nin': ['cancelled', 'failed']}
            })
            
            print(f"🎫 {code}")
            print(f"   Max Uses: {max_uses}")
            print(f"   Stored Count: {usage_count}")
            print(f"   Real Count: {real_count}")
            
            if max_uses is not None and max_uses > 0:
                if real_count >= max_uses:
                    print(f"   ❌ SHOULD BE BLOCKED (Real: {real_count}/{max_uses})")
                else:
                    print(f"   ✅ Should be valid (Real: {real_count}/{max_uses})")
            else:
                print(f"   ♾️  Unlimited usage")
            print()
        
        # Test API validation for existing coupons
        print("\n🔬 TESTING API VALIDATION:")
        print("-" * 30)
        
        backend_url = "http://localhost:8000"  # Adjust if needed
        
        for coupon in active_coupons[:3]:  # Test first 3 coupons
            code = coupon.get('code')
            max_uses = coupon.get('maxUses')
            
            # Count real orders
            real_count = await orders_collection.count_documents({
                '$or': [
                    {'couponCode': {'$regex': f'^{code}$', '$options': 'i'}},
                    {'coupon_code': {'$regex': f'^{code}$', '$options': 'i'}}
                ],
                'status': {'$nin': ['cancelled', 'failed']}
            })
            
            # Determine expected result
            should_be_blocked = False
            if max_uses is not None and max_uses > 0 and real_count >= max_uses:
                should_be_blocked = True
            
            print(f"Testing coupon: {code}")
            print(f"Real usage: {real_count}, Max: {max_uses}")
            print(f"Expected: {'BLOCKED' if should_be_blocked else 'VALID'}")
            
            try:
                # Test API validation
                response = requests.post(f"{backend_url}/api/coupons/validate", 
                    json={
                        "code": code,
                        "amount": 100,
                        "email": "test@example.com"
                    },
                    timeout=10
                )
                
                if response.status_code == 200:
                    data = response.json()
                    is_valid = data.get('valid', True) and not data.get('error') and data.get('discount', 0) > 0
                    
                    print(f"API Response: {data}")
                    
                    if should_be_blocked:
                        if is_valid:
                            print("❌ FAIL: Coupon should be blocked but API says it's valid!")
                        else:
                            print("✅ PASS: Coupon correctly blocked")
                    else:
                        if is_valid:
                            print("✅ PASS: Coupon correctly valid")
                        else:
                            print("❌ FAIL: Coupon should be valid but API says it's blocked!")
                else:
                    print(f"⚠️ API Error: HTTP {response.status_code}")
                    
            except requests.exceptions.RequestException as e:
                print(f"⚠️ Cannot connect to backend: {e}")
                print("   (Make sure the backend server is running)")
            
            print("-" * 20)
        
        client.close()
        
    except Exception as e:
        logger.error(f"Error during test: {e}")
        import traceback
        traceback.print_exc()

if __name__ == "__main__":
    asyncio.run(test_coupon_max_usage_fix())

File: test_max_usage_fix.js
// Test script to verify the coupon max usage fix
const axios = require('axios');

const API_URL = 'https://api.monkeyz.co.il';

async function testCouponMaxUsageFix() {
    console.log('🔧 Testing Coupon Max Usage Fix...\n');
    
    // Test configuration - replace with actual values from your system
    const testCoupon = 'TEST10'; // Replace with a real coupon code
    const testEmail1 = 'test1@example.com';
    const testEmail2 = 'test2@example.com';
    const testAmount = 100;
    
    console.log('📋 Test Scenario:');
    console.log(`- Coupon: ${testCoupon}`);
    console.log('- Should have maxUsagePerUser: 1 and maxUses: 2');
    console.log('- Expected behavior: Work max 2 times total, max 1 time per user\n');
    
    try {
        console.log('Step 1: Validate coupon with first email (should work)...');
        const validate1 = await axios.post(`${API_URL}/api/coupons/validate`, {
            code: testCoupon,
            amount: testAmount,
            email: testEmail1
        });
        
        console.log(`✅ Validation 1: Valid=${validate1.data.valid}, Discount=${validate1.data.discount}`);
        
        console.log('\nStep 2: Validate coupon with second email (should work)...');
        const validate2 = await axios.post(`${API_URL}/api/coupons/validate`, {
            code: testCoupon,
            amount: testAmount,
            email: testEmail2
        });
        
        console.log(`✅ Validation 2: Valid=${validate2.data.valid}, Discount=${validate2.data.discount}`);
        
        // This test only validates - it doesn't actually apply the coupon
        // The real test is when users actually complete orders through PayPal
        
        console.log('\n🎯 What the fix addresses:');
        console.log('✅ create_paypal_order() now uses apply_coupon() instead of validate_coupon()');
        console.log('✅ This increments usage count immediately when PayPal order is created');
        console.log('✅ capture_paypal_order() no longer double-applies (just updates analytics)');
        console.log('✅ Prevents race condition where multiple orders could be created before usage incremented');
        
        console.log('\n📝 Technical Changes Made:');
        console.log('1. In create_paypal_order(): Changed validate_coupon() → apply_coupon()');
        console.log('2. In capture_paypal_order(): Removed coupon application, only updates analytics');
        console.log('3. Now usage count increments at order creation, not payment capture');
        console.log('4. Frontend email change fix prevents stale discount display');
        
        console.log('\n⚠️  Manual Testing Required:');
        console.log('1. Create a coupon with maxUsagePerUser=1, maxUses=2');
        console.log('2. Try to use it with same email twice → should fail on second attempt');
        console.log('3. Try to use it 3 times total → should fail on third attempt');
        console.log('4. Verify usage count increments immediately when PayPal order created');
        
    } catch (error) {
        console.error('❌ Test failed:', error.message);
        if (error.response?.data) {
            console.error('Error details:', error.response.data);
        }
    }
}

console.log('🧪 Coupon Max Usage Bug Fix Test\n');
console.log('This test verifies the fix for the issue where coupons with');
console.log('maxUsagePerUser=1 and maxUses=2 were working more than twice.\n');

testCouponMaxUsageFix();
